File: hadwiger_nelson_overlapping_forcing_seed/colour_check.cpp
// Native execution of the documented exhaustive four-colour search.
#include <algorithm>
#include <array>
#include <cstdint>
#include <iostream>
#include <limits>
#include <stdexcept>
#include <vector>
using Mask=unsigned;
struct Search {
    int n; std::vector<std::vector<int>> adj;
    std::uint64_t nodes=0, conflicts=0;
    std::vector<Mask> solution;
    static unsigned count(Mask m) { return unsigned(__builtin_popcount(m)); }
    static void increment(std::uint64_t& x) {
        if(x==std::numeric_limits<std::uint64_t>::max())throw std::runtime_error("Counter overflow");
        ++x;
    }
    bool dfs(std::vector<Mask> masks,std::vector<int> queue) {
        increment(nodes);
        while(!queue.empty()) {
            int v=queue.back();queue.pop_back();Mask bit=masks[v];
            if(bit==0 || (bit&(bit-1)))throw std::runtime_error("Bad propagation queue");
            for(int u:adj[v])if(masks[u]&bit) {
                Mask next=masks[u]&~bit;
                if(!next){increment(conflicts);return false;}
                masks[u]=next;
                if(!(next&(next-1)))queue.push_back(u);
            }
        }
        int best=-1;unsigned size=5;Mask used=0;
        for(int v=0;v<n;++v) {
            Mask m=masks[v];
            if(!(m&(m-1))){used|=m;continue;}
            unsigned sz=count(m);
            if(sz<size || (sz==size && (best<0 || adj[v].size()>adj[best].size()))) {
                best=v;size=sz;
            }
        }
        if(best<0){solution=std::move(masks);return true;}
        Mask candidates=masks[best]&used,unused=masks[best]&~used;
        if(unused)candidates|=unused&(~unused+1U);
        while(candidates) {
            Mask bit=candidates&(~candidates+1U);candidates-=bit;
            auto child=masks;child[best]=bit;
            if(dfs(std::move(child),{best}))return true;
        }
        return false;
    }
};
int main() {
    try {
        int n,m,p;
        if(!(std::cin>>n>>m>>p) || n<0 || n>4096 || m<0 || p<0 || p>n ||
           static_cast<long long>(m)>static_cast<long long>(n)*(n-1)/2)
            throw std::runtime_error("Bad dimensions");
        Search s{n,std::vector<std::vector<int>>(n),0,0,{}};
        for(int i=0;i<m;++i) {
            int a,b;if(!(std::cin>>a>>b)||a<0||a>=n||b<0||b>=n||a==b)
                throw std::runtime_error("Bad edge");
            s.adj[a].push_back(b);s.adj[b].push_back(a);
        }
        for(auto& row:s.adj) {
            std::sort(row.begin(),row.end());
            if(std::adjacent_find(row.begin(),row.end())!=row.end())throw std::runtime_error("Duplicate edge");
        }
        std::vector<Mask> masks(n,15);std::vector<int> queue;
        for(int i=0;i<p;++i) {
            int v,c;if(!(std::cin>>v>>c)||v<0||v>=n||c<0||c>3||masks[v]!=15)
                throw std::runtime_error("Bad pin");
            masks[v]=1U<<unsigned(c);queue.push_back(v);
        }
        std::string tail;if(std::cin>>tail)throw std::runtime_error("Trailing data");
        bool sat=s.dfs(std::move(masks),std::move(queue));
        std::cout<<"{\"satisfiable\":"<<(sat?"true":"false")<<",\"nodes\":"<<s.nodes
                 <<",\"conflicts\":"<<s.conflicts<<",\"colouring\":[";
        if(sat)for(int v=0;v<n;++v) {
            if(v)std::cout<<',';
            std::cout<<unsigned(__builtin_ctz(s.solution[v]));
        }
        std::cout<<"]}\n";
        return 0;
    }catch(const std::exception& e){std::cerr<<e.what()<<'\n';return 2;}
}

File: hadwiger_nelson_small_triangle_forcer375/colour_check.py
"""Exact exhaustive four-colour search with singleton propagation."""
def solve(n, edges, pins=(), triples=()):
    adj = [set() for _ in range(n)]
    incident = [[] for _ in range(n)]
    for u,v in edges:
        adj[u].add(v);adj[v].add(u)
    for t in triples:
        for v in t:incident[v].append(t)
    adj=[sorted(a) for a in adj]
    nodes=0;conflicts=0
    def propagate(masks,queue):
        while queue:
            v=queue.pop();bit=masks[v]
            if not bit or bit & (bit-1):raise RuntimeError('Non-singleton queue')
            for u in adj[v]:
                old=masks[u]
                if old & bit:
                    new=old & ~bit
                    if not new:return False
                    masks[u]=new
                    if new & (new-1)==0:queue.append(u)
            for t in incident[v]:
                fixed=[u for u in t if masks[u]==bit]
                if len(fixed)>=2:
                    for u in t:
                        if u in fixed:continue
                        old=masks[u];new=old & ~bit
                        if not new:return False
                        if new!=old:
                            masks[u]=new
                            if new & (new-1)==0:queue.append(u)
                    if len(fixed)==3:return False
        return True
    def dfs(masks,queue):
        nonlocal nodes,conflicts
        nodes+=1
        if not propagate(masks,queue):conflicts+=1;return None
        choices=[v for v,m in enumerate(masks) if m & (m-1)]
        if not choices:return [m.bit_length()-1 for m in masks]
        v=min(choices,key=lambda v:(masks[v].bit_count(),-len(adj[v]),v))
        used=0
        for m in masks:
            if m & (m-1)==0:used|=m
        candidates=masks[v] & used
        unused=masks[v] & ~used
        if unused:candidates |= unused & -unused
        while candidates:
            bit=candidates & -candidates;candidates-=bit
            child=masks.copy();child[v]=bit
            answer=dfs(child,[v])
            if answer is not None:return answer
        return None
    masks=[15]*n
    for v,c in pins:
        if masks[v]!=15 and masks[v]!=(1<<c):return None,{'nodes':0,'conflicts':1}
        masks[v]=1<<c
    answer=dfs(masks,[v for v,c in pins])
    return answer,{'nodes':nodes,'conflicts':conflicts}
